genkit-tools/common/src/utils/utils.ts

```diff
@@ -215,3 +223,57 @@ export async function retriable<T>(
     attempt++;
   }
 }
+
+/**
+ * Checks if the provided data is a valid dev tools server state file.
+ */
+export function isValidDevToolsInfo(data: any): data is DevToolsInfo {
+  return (
+    typeof data === 'object' &&
+    typeof data.url === 'string' &&
+    typeof data.timestamp === 'string'
+  );
+}
+
+/**
+ * Fetches the Genkit Dev UI URL if available
+ */
+export async function getDevUiUrl(
+  projectRoot?: string
+): Promise<string | undefined> {
+  const serversDir = await findServersDir(projectRoot);
+  const toolsJsonPath = path.join(serversDir, 'tools.json');
```

Comment:
you're always writing to `tools.json`? this assumes we track only the last instance of the dev ui? do we want to track all instances?

Reply:
I did not know that we'd have multiple Dev UI instances on the same runtime. What would the difference between the instances be?

I can add support for multiple instances if we want, I just need some direction on how we do that for other state files.

Reply:
I assume like if you're working on a local version of the Dev UI.

Reply:
I dont think we can handle supporting a local Dev UI, as it is a separate process entirely and not tracked through `genkit start`

Reply:
I added support for multiple Dev UI servers, PTAL.